From f863c32908eb53ce6805d387739c82bbbec3612f Mon Sep 17 00:00:00 2001
From: AfalpHy <736353503@qq.com>
Date: Wed, 12 Feb 2025 14:53:48 +0800
Subject: [PATCH 02/29] fix psram addr bug

---
 perip/psram/psram.v | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/perip/psram/psram.v b/perip/psram/psram.v
index 4b63f7f2..f2ec0124 100644
--- a/perip/psram/psram.v
+++ b/perip/psram/psram.v
@@ -23,7 +23,7 @@ module psram (
   reg [7:0] cnt;
 
   reg [7:0] cmd;
-  reg [31:0] addr;
+  reg [23:0] addr;
   wire [31:0] rdata;
   reg [31:0] data;
 
@@ -32,7 +32,7 @@ module psram (
   assign dio = (~wen & state == DATA) ? data[7:4] : 4'bz;
 
   always @(state) begin
-    if (state == DATA && ~wen) psram_read(addr, rdata);
+    if (state == DATA && ~wen) psram_read({8'b0, addr}, rdata);
   end
 
   always @(posedge sck or posedge ce_n) begin
@@ -65,7 +65,7 @@ module psram (
   always @(posedge sck) begin
     case (state)
       CMD: cmd <= {cmd[6:0], dio[0]};
-      ADDR: addr <= {addr[27:0], dio};
+      ADDR: addr <= {addr[19:0], dio};
       DATA: begin
         if (wen) data <= {data[27:0], dio};
         else begin
@@ -85,11 +85,11 @@ module psram (
   always @(posedge ce_n) begin
     if (wen) begin
       if (cnt == 16) begin
-        psram_write(addr, {24'b0, data[7:0]}, 32'hff);
+        psram_write({8'b0, addr}, {24'b0, data[7:0]}, 32'hff);
       end else if (cnt == 18) begin
-        psram_write(addr, {16'b0, data[7:0], data[15:8]}, 32'hffff);
+        psram_write({8'b0, addr}, {16'b0, data[7:0], data[15:8]}, 32'hffff);
       end else if (cnt == 22 && wen) begin
-        psram_write(addr, {data[7:0], data[15:8], data[23:16], data[31:24]}, 32'hffff_ffff);
+        psram_write({8'b0, addr}, {data[7:0], data[15:8], data[23:16], data[31:24]}, 32'hffff_ffff);
       end
     end
   end
-- 
2.34.1

