From 400975eef1f65c37944497a2ea551256a302389e Mon Sep 17 00:00:00 2001
From: AfalpHy <736353503@qq.com>
Date: Sun, 16 Feb 2025 11:16:30 +0800
Subject: [PATCH 09/29] fix sdram read signal bug

---
 perip/sdram/sdram.v | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/perip/sdram/sdram.v b/perip/sdram/sdram.v
index c32fb040..22d70423 100644
--- a/perip/sdram/sdram.v
+++ b/perip/sdram/sdram.v
@@ -40,6 +40,8 @@ module sdram (
   reg [1:0] dqm_buffer;
 
   reg read;
+  reg burst;
+  reg [1:0] burst_len;
   assign dq = read ? data_buffer[15:0] : 16'hz;
 
   reg [23:0] addr;
@@ -58,7 +60,6 @@ module sdram (
           state <= ACTIVATE;
         end
         CMD_READ: begin
-          read <= 1;
           addr[8:1] <= a[8:1];
           addr[10:9] <= ba;
           state <= READ;
@@ -70,6 +71,11 @@ module sdram (
           dqm_buffer <= dqm;
           state <= WRITE0;
         end
+        CMD_LOAD_MODE: begin
+          burst <= a[12:10] == 3'b0;
+          burst_len <= a[6:5];
+          state <= LOAD_MODE;
+        end
         default: state <= state;
       endcase
     end
@@ -77,6 +83,8 @@ module sdram (
 
   always @(posedge clk) begin
     if (cke) begin
+      if (state == READ || state == READ_WAIT) read <= 1;
+      else read <= 0;
       if (state == READ) begin
         data_buffer <= {memory[addr+1], memory[addr]};
       end else if (state == READ_WAIT) begin
-- 
2.34.1

