From 077e04526e90bbf9cb70b6108f4dad333f050d35 Mon Sep 17 00:00:00 2001
From: AfalpHy <736353503@qq.com>
Date: Mon, 17 Feb 2025 11:04:48 +0800
Subject: [PATCH 14/29] sdram word extend

---
 perip/sdram/sdram_top_apb.v | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/perip/sdram/sdram_top_apb.v b/perip/sdram/sdram_top_apb.v
index 781bcc13..5b0cfe94 100644
--- a/perip/sdram/sdram_top_apb.v
+++ b/perip/sdram/sdram_top_apb.v
@@ -21,9 +21,11 @@ module sdram_top_apb (
   output [12:0] sdram_a,
   output [ 1:0] sdram_ba,
   output [ 3:0] sdram_dqm,
-  inout  [31:0] sdram_dq
+  inout  [31:0] sdram_dq,
+  output        sdram_sel
 );
 
+  assign sdram_sel = in_paddr[25];
   wire sdram_dout_en;
   wire [31:0] sdram_dout;
   assign sdram_dq = sdram_dout_en ? sdram_dout : 32'bz;
-- 
2.34.1

