From 6c3da7ccbf1600369315adea8c192917c2ae9aca Mon Sep 17 00:00:00 2001
From: AfalpHy <736353503@qq.com>
Date: Mon, 17 Feb 2025 15:13:50 +0800
Subject: [PATCH 16/29] process gpio_in

---
 perip/gpio/gpio_top_apb.v | 1 +
 1 file changed, 1 insertion(+)

diff --git a/perip/gpio/gpio_top_apb.v b/perip/gpio/gpio_top_apb.v
index eacbbdf8..ff805bcd 100644
--- a/perip/gpio/gpio_top_apb.v
+++ b/perip/gpio/gpio_top_apb.v
@@ -32,6 +32,7 @@ module gpio_top_apb (
 
   assign gpio_out  = led_data;
   assign in_pready = ack;
+  assign in_prdata = {16'b0, gpio_in};
 
   always @(posedge clock) begin
     if (reset) begin
-- 
2.34.1

