From f69930ac16f7d7db871c8a0bf6aeb52ef184d700 Mon Sep 17 00:00:00 2001
From: AfalpHy <736353503@qq.com>
Date: Tue, 18 Feb 2025 10:05:22 +0800
Subject: [PATCH 17/29] reduce sdram size

---
 perip/sdram/sdram.v | 16 ++++++++--------
 1 file changed, 8 insertions(+), 8 deletions(-)

diff --git a/perip/sdram/sdram.v b/perip/sdram/sdram.v
index aeca4dae..b7be1028 100644
--- a/perip/sdram/sdram.v
+++ b/perip/sdram/sdram.v
@@ -32,7 +32,7 @@ module sdram (
 
   reg [3:0] state;
 
-  reg [15:0] memory[0:'h1000000-1];
+  reg [15:0] memory[0:'h800000-1];
 
   reg [15:0] data_buffer;
   reg [15:0] dqm_buffer;
@@ -42,7 +42,7 @@ module sdram (
   reg [1:0] burst_len;
   assign dq = read ? data_buffer : 16'hz;
 
-  reg [23:0] addr;
+  reg [22:0] addr;
   reg [12:0] addr_row[0:3];
 
   always @(posedge clk) begin
@@ -56,15 +56,15 @@ module sdram (
           state <= ACTIVATE;
         end
         CMD_READ: begin
-          addr[8:1] <= a[8:1];
-          addr[10:9] <= ba;
-          addr[23:11] <= addr_row[ba];
+          addr[7:0] <= a[8:1];
+          addr[9:8] <= ba;
+          addr[22:10] <= addr_row[ba];
           state <= READ;
         end
         CMD_WRITE: begin
-          addr[8:1] <= a[8:1];
-          addr[10:9] <= ba;
-          addr[23:11] <= addr_row[ba];
+          addr[7:0] <= a[8:1];
+          addr[9:8] <= ba;
+          addr[22:10] <= addr_row[ba];
           data_buffer[15:0] <= dq;
           dqm_buffer <= {{8{dqm[1]}}, {8{dqm[0]}}};
           state <= WRITE0;
-- 
2.34.1

