From 87dd36dc4088495c5301e01c0e99980a68995583 Mon Sep 17 00:00:00 2001
From: AfalpHy <736353503@qq.com>
Date: Tue, 18 Feb 2025 10:31:40 +0800
Subject: [PATCH 18/29] reduce gpio delay

---
 perip/gpio/gpio_top_apb.v | 8 +++-----
 1 file changed, 3 insertions(+), 5 deletions(-)

diff --git a/perip/gpio/gpio_top_apb.v b/perip/gpio/gpio_top_apb.v
index ff805bcd..37a66084 100644
--- a/perip/gpio/gpio_top_apb.v
+++ b/perip/gpio/gpio_top_apb.v
@@ -31,22 +31,20 @@ module gpio_top_apb (
   wire [31:0] mask = {{8{in_pstrb[3]}}, {8{in_pstrb[2]}}, {8{in_pstrb[1]}}, {8{in_pstrb[0]}}};
 
   assign gpio_out  = led_data;
-  assign in_pready = ack;
+  assign in_pready = in_psel & in_penable;
   assign in_prdata = {16'b0, gpio_in};
 
   always @(posedge clock) begin
     if (reset) begin
       led_data <= 0;
       bcd_data <= 0;
-      ack <= 0;
     end else begin
-      if (in_psel && in_penable) begin
+      if (in_psel && ~in_penable) begin
         if (in_pwrite) begin
           if (in_paddr[3]) bcd_data <= (bcd_data & ~mask) | (in_pwdata & mask);
           else led_data <= (led_data & ~(mask[15:0]) | (in_pwdata[15:0] & mask[15:0]));
         end
-        ack <= 1;
-      end else ack <= 0;
+      end
     end
   end
 
-- 
2.34.1

