From b6bbd6afdef2f773b8be893b83db797caf1d3e9b Mon Sep 17 00:00:00 2001
From: AfalpHy <736353503@qq.com>
Date: Tue, 18 Feb 2025 14:27:45 +0800
Subject: [PATCH 20/29] add comment

---
 perip/ps2/ps2_top_apb.v | 1 +
 1 file changed, 1 insertion(+)

diff --git a/perip/ps2/ps2_top_apb.v b/perip/ps2/ps2_top_apb.v
index 08dd3c24..39e53793 100644
--- a/perip/ps2/ps2_top_apb.v
+++ b/perip/ps2/ps2_top_apb.v
@@ -69,6 +69,7 @@ module ps2_top_apb (
               key_buffer[0] <= buffer[8:1];
               key_len <= key_len + 1;
             end else begin
+              // Ensure that the complete key code is written, discard when it can't be store
               if (key_len == 2 && (w_ptr + 3 != r_ptr)) begin
                 fifo[w_ptr] <= key_buffer[1];
                 fifo[w_ptr+1] <= key_buffer[0];
-- 
2.34.1

