From 334d5a6fb91263f5681f0400d3fcc645f3ba1d3b Mon Sep 17 00:00:00 2001
From: AfalpHy <736353503@qq.com>
Date: Sat, 22 Feb 2025 16:31:04 +0800
Subject: [PATCH 24/29] trivial

---
 perip/vga/vga_top_apb.v | 20 ++++++++++----------
 1 file changed, 10 insertions(+), 10 deletions(-)

diff --git a/perip/vga/vga_top_apb.v b/perip/vga/vga_top_apb.v
index 3945a19c..9c7fe871 100644
--- a/perip/vga/vga_top_apb.v
+++ b/perip/vga/vga_top_apb.v
@@ -26,25 +26,25 @@ module vga_top_apb (
 
   reg [20:0] addr;
 
-  reg state;
+  reg update;
 
   assign vga_r = buffer[addr+2];
   assign vga_g = buffer[addr+1];
   assign vga_b = buffer[addr];
 
-  assign vga_valid = state;
+  assign vga_valid = update;
 
   always @(posedge clock) begin
     if (reset) begin
-      addr  <= 0;
-      state <= 0;
+      addr   <= 0;
+      update <= 0;
     end else begin
       if (in_psel & ~in_penable) begin
         if (in_pwrite) begin
           if (in_paddr[20:0] == 'h1ffffc) begin
-            if (in_pwdata == 'h1 && state == 0) begin
-              state <= 1;
-              addr  <= 0;
+            if (in_pwdata == 'h1 && update == 0) begin
+              update <= 1;
+              addr   <= 0;
             end
           end else begin
             if (in_pstrb[0]) buffer[in_paddr[20:0]] <= in_pwdata[7:0];
@@ -54,10 +54,10 @@ module vga_top_apb (
           end
         end
       end
-      if (state) begin
+      if (update) begin
         if (addr + 4 == 640 * 480 * 4) begin
-          state <= 0;
-          addr  <= 0;
+          update <= 0;
+          addr   <= 0;
         end else addr <= addr + 4;
       end
     end
-- 
2.34.1

