From e96b813db7677b814b9f9dc9045c8970de5fdfc6 Mon Sep 17 00:00:00 2001
From: AfalpHy <736353503@qq.com>
Date: Sun, 16 Mar 2025 11:53:49 +0800
Subject: [PATCH 26/29] update delay

---
 perip/amba/apb_delayer.v | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/perip/amba/apb_delayer.v b/perip/amba/apb_delayer.v
index 48fad4e7..f2d7fdd7 100644
--- a/perip/amba/apb_delayer.v
+++ b/perip/amba/apb_delayer.v
@@ -37,11 +37,11 @@ module apb_delayer (
     end else begin
       if (pready) pready <= 0;
       else if (in_penable & out_pready) begin
-        counter = counter + 360;  // 2.8 * 128(cpu clock 380MHZ)
+        counter = counter + 940;  // 7.4 * 128(cpu clock 380MHZ)
         counter = counter >> 7;
         done   <= 1;
         prdata <= out_prdata;
-      end else if (out_penable) counter = counter + 360;
+      end else if (out_penable) counter = counter + 940;
       else if (in_penable) begin
         counter = counter - 1;
         if (counter == 0) begin
-- 
2.34.1

