From 84a3810b31ab52061c973ea8c52c10231b7d0ed0 Mon Sep 17 00:00:00 2001
From: AfalpHy <736353503@qq.com>
Date: Sat, 2 Aug 2025 22:37:54 +0800
Subject: [PATCH] avoid loops in testing

---
 bsp/abstract-machine/src/uart.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/bsp/abstract-machine/src/uart.c b/bsp/abstract-machine/src/uart.c
index 69af3fb8c..443ff80fa 100644
--- a/bsp/abstract-machine/src/uart.c
+++ b/bsp/abstract-machine/src/uart.c
@@ -39,7 +39,10 @@ static int _uart_putc(struct rt_serial_device *serial, char c) {
 
 static int _uart_getc(struct rt_serial_device *serial) {
   static const char *p = "help\ndate\nversion\nfree\nps\npwd\nls\nmemtrace\nmemcheck\nutest_list\n";
-  return (*p != '\0' ? *(p ++) : io_read(AM_UART_RX).data);
+  if (*p != '\0') {
+    return *(p++);
+  }
+  halt(0);
 }
 
 const struct rt_uart_ops _uart_ops = {
-- 
2.34.1

